File: benchmarks/pandas/bench_get_dummies_opts.py
"""
Benchmark: pandas.get_dummies with prefix, drop_first, dummy_na options.
Outputs JSON: {"function": "get_dummies_opts", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import pandas as pd

SIZE = 10_000
WARMUP = 5
ITERATIONS = 30

categories = ["apple", "banana", "cherry", "date", "elderberry"]
data = [None if i % 20 == 0 else categories[i % len(categories)] for i in range(SIZE)]
s = pd.Series(data, dtype="object")

df = pd.DataFrame({
    "fruit": [None if i % 20 == 0 else categories[i % len(categories)] for i in range(SIZE)],
    "color": [["red", "green", "blue"][i % 3] for i in range(SIZE)],
})

for _ in range(WARMUP):
    pd.get_dummies(s, prefix="cat", dummy_na=True)
    pd.get_dummies(s, drop_first=True)
    pd.get_dummies(df, columns=["fruit", "color"], prefix="col", drop_first=True)

start = time.perf_counter()
for _ in range(ITERATIONS):
    pd.get_dummies(s, prefix="cat", dummy_na=True)
    pd.get_dummies(s, drop_first=True)
    pd.get_dummies(df, columns=["fruit", "color"], prefix="col", drop_first=True)
total = (time.perf_counter() - start) * 1000

print(json.dumps({"function": "get_dummies_opts", "mean_ms": total / ITERATIONS, "iterations": ITERATIONS, "total_ms": total}))

File: benchmarks/tsb/bench_get_dummies_opts.ts
/**
 * Benchmark: getDummies / dataFrameGetDummies with prefix, dropFirst, dummyNa options.
 * Outputs JSON: {"function": "get_dummies_opts", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { getDummies, dataFrameGetDummies, Series, DataFrame } from "../../src/index.ts";

const SIZE = 10_000;
const WARMUP = 5;
const ITERATIONS = 30;

const categories = ["apple", "banana", "cherry", "date", "elderberry"];
const data = Array.from({ length: SIZE }, (_, i) =>
  i % 20 === 0 ? null : categories[i % categories.length],
);
const s = new Series({ data });

const df = DataFrame.fromColumns({
  fruit: Array.from({ length: SIZE }, (_, i) =>
    i % 20 === 0 ? null : categories[i % categories.length],
  ),
  color: Array.from({ length: SIZE }, (_, i) => ["red", "green", "blue"][i % 3]),
});

for (let i = 0; i < WARMUP; i++) {
  getDummies(s, { prefix: "cat", dummyNa: true });
  getDummies(s, { dropFirst: true });
  dataFrameGetDummies(df, { columns: ["fruit", "color"], prefix: "col", dropFirst: true });
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  getDummies(s, { prefix: "cat", dummyNa: true });
  getDummies(s, { dropFirst: true });
  dataFrameGetDummies(df, { columns: ["fruit", "color"], prefix: "col", dropFirst: true });
}
const total = performance.now() - start;

console.log(JSON.stringify({ function: "get_dummies_opts", mean_ms: total / ITERATIONS, iterations: ITERATIONS, total_ms: total }));
